### PR TITLE
Ensure that the lock screen is maintained when the app is supposed to be locked.

### LINKED ISSUE #260: When in locked state app can re-open to list view
During demo by @jhugman we observed the app being manually "Locked" and then backgrounded.

Upon re-opening, it opened the entry list view (not the locked/unlock button view).

### PR DESCRIPTION
**Pull request by jhugman**
There are serveral parts to this, each of which is slow and difficult. 

 - [x] Ensure that the AutoLockTests work.
 - [x] Write failing new tests for the Locking edge cases
 - [x] Correct the application code such that the new tests and `AutoLockTests` all work.

This PR: 
 * updates the use of the android support library to androidx.
 * stabilizes a lot of tests by moving to the AndroidX Test Coordinator. 
 * re-writes `AutoLockStore` to broadcast the need for locking, rather than autolocking would fire here if it needed here. This makes auto locking a little more encapsulated.
 * `AutoLockStoreTest` needed changing in order to make that work.
 * It also moves a lot of logic out of the `RoutePresenter` to the various stores.
    - `DataStore`
    - `RouteStore`
    - `AutoLockStore`

Fixes #260

### CODE REVIEW

File: app/src/main/java/mozilla/lockbox/view/DialogFragment.kt [hunk 1]
```diff
@@ -7,12 +7,12 @@
...
 import android.os.Bundle
-import android.support.annotation.StringRes
-import android.support.v4.app.DialogFragment
+import androidx.annotation.StringRes
+import androidx.fragment.app.DialogFragment
 import android.view.View
 import mozilla.lockbox.flux.Presenter
 
-open class DialogFragment : DialogFragment() {
+open class DialogFragment : androidx.fragment.app.DialogFragment() {
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/view/DialogFragment.kt]: nit: import this rather than having the long class name

File: app/src/main/java/mozilla/lockbox/view/FilterFragment.kt [hunk 2]
```diff
@@ -38,10 +38,10 @@ class FilterFragment : BackableFragment(), FilterView {
         presenter = FilterPresenter(this)
         val view = inflater.inflate(R.layout.fragment_filter, container, false)
 
-        val layoutManager = LinearLayoutManager(context)
+        val layoutManager = androidx.recyclerview.widget.LinearLayoutManager(context)
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/view/FilterFragment.kt]: nit: import over long class name

File: app/src/main/res/layout/fragment_fingerprint_dialog.xml [hunk 3]
```diff
@@ -4,15 +4,15 @@
   ~ file, You can obtain one at http://mozilla.org/MPL/2.0/.
   -->
 
-<android.support.constraint.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
-    xmlns:app="http://schemas.android.com/apk/res-auto"
-    xmlns:tools="http://schemas.android.com/tools"
-    android:layout_width="wrap_content"
-    android:layout_height="wrap_content"
-    android:paddingTop="32dp">
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (sashei) [inline app/src/main/res/layout/fragment_fingerprint_dialog.xml]: 👾

File: app/src/main/res/layout/fragment_item_list.xml [hunk 4]
```diff
@@ -4,21 +4,21 @@
...
   -->
-<android.support.v4.widget.DrawerLayout xmlns:android="http://schemas.android.com/apk/res/android"
-    xmlns:app="http://schemas.android.com/apk/res-auto"
-    xmlns:tools="http://schemas.android.com/tools"
-    android:id="@+id/appDrawer"
-    android:layout_width="match_parent"
-    android:layout_height="match_parent"
-    tools:context=".view.ItemListFragment">
+<androidx.drawerlayout.widget.DrawerLayout xmlns:android="http://schemas.android.com/apk/res/android"
+                                           xmlns:app="http://schemas.android.com/apk/res-auto"
```
Comment (sashei) [inline app/src/main/res/layout/fragment_item_list.xml]: 👾

File: app/src/androidTest/java/mozilla/lockbox/Navigator.kt [hunk 5]
```diff
@@ -38,14 +42,46 @@ class Navigator {
...
+                Dispatcher.shared.dispatch(DataStoreAction.Unlock)
+                blockUntil(DataStore.State.Unlocked)
+            }
+
+            is DataStore.State.Unprepared -> {
+                return
+            }
+
+            else -> {
+                // NOP
```
Comment (eliserichards) [inline app/src/androidTest/java/mozilla/lockbox/Navigator.kt]: Should this be empty?

File: app/src/androidTest/java/mozilla/lockbox/RoutePresenterTest.kt [hunk 6]
```diff
@@ -30,11 +29,6 @@ open class RoutePresenterTest {
     @Rule @JvmField
     val activityRule: ActivityTestRule<RootActivity> = ActivityTestRule(RootActivity::class.java)
 
-    @Before
-    fun setUp() {
-        navigator.resetApp(activityRule)
```
Comment (eliserichards) [inline app/src/androidTest/java/mozilla/lockbox/RoutePresenterTest.kt]: Why do we no longer have to use `resetApp`?

**Comment by eliserichards** [comment]
LGTM! 

I see some tests failing on the branch locally (`SettingListAdapterTest`, `SectionedAdapterTest`), but I assume those are the reason that bitrise is red.

File: app/src/androidTest/java/mozilla/lockbox/RoutePresenterTest.kt (hunk 6 above)
Comment (jhugman) [reply app/src/androidTest/java/mozilla/lockbox/RoutePresenterTest.kt]: Moving over to the test coordinator means we're now clearing the data in the app before each test: in essence doing a better job of resetting the app that we were doing here.

Result: more stable tests!

File: app/src/main/java/mozilla/lockbox/view/DialogFragment.kt (hunk 1 above)
Comment (jhugman) [reply app/src/main/java/mozilla/lockbox/view/DialogFragment.kt]: :+1:

File: app/src/androidTest/java/mozilla/lockbox/Navigator.kt (hunk 5 above)
Comment (jhugman) [reply app/src/androidTest/java/mozilla/lockbox/Navigator.kt]: Yes. The `// NOP` is short for noop. We should probably make this comment more like '// do nothing`, minimally.
